{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3b1c28a6",
   "metadata": {},
   "source": [
    "# ðŸ§¬ Pan-Cancer Immune Biomarker Discovery (with Real Survival Data)\n",
    "\n",
    "This notebook integrates TCGA expression, immune subtypes, and real clinical survival metadata from cBioPortal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20762c89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: Could not find a version that satisfies the requirement gdc-client (from versions: none)\u001b[0m\u001b[31m\n",
      "\u001b[0m\u001b[31mERROR: No matching distribution found for gdc-client\u001b[0m\u001b[31m\n",
      "\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m24.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install -q requests pandas seaborn matplotlib lifelines scikit-learn gdown"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4a1ec0c",
   "metadata": {},
   "source": [
    "## ðŸ”Œ Connect to cBioPortal API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "dba1517c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "# Base URL for the API\n",
    "base_url = 'https://www.cbioportal.org/api'\n",
    "\n",
    "def get_cbioportal_data(endpoint, params=None):\n",
    "    url = f\"{base_url}/{endpoint}\"\n",
    "    response = requests.get(url, params=params)\n",
    "    response.raise_for_status()  # Raise an exception for bad status codes\n",
    "    return response.json()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62ecc824",
   "metadata": {},
   "source": [
    "## ðŸ“¦ Get TCGA BRCA Sample List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9d2a06cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Available studies:\n",
      "Study ID: all_stjude_2015, Name: Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2015)\n",
      "Study ID: all_stjude_2013, Name: Hypodiploid Acute Lymphoid Leukemia (St Jude, Nat Genet 2013)\n",
      "Study ID: acyc_fmi_2014, Name: Adenoid Cystic Carcinoma (FMI, Am J Surg Pathl. 2014)\n",
      "Study ID: acyc_jhu_2016, Name: Adenoid Cystic Carcinoma (JHU, Cancer Prev Res 2016)\n",
      "Study ID: acyc_mda_2015, Name: Adenoid Cystic Carcinoma (MDA, Clin Cancer Res 2015)\n",
      "Study ID: acyc_mgh_2016, Name: Adenoid Cystic Carcinoma (MGH, Nat Gen 2016)\n",
      "Study ID: acyc_sanger_2013, Name: Adenoid Cystic Carcinoma (Sanger/MDA, JCI 2013)\n",
      "Study ID: all_stjude_2016, Name: Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2016)\n",
      "Study ID: appendiceal_msk_2022, Name: Appendiceal Cancer (MSK, J Clin Oncol 2022)\n",
      "Study ID: blca_plasmacytoid_mskcc_2016, Name: Bladder Cancer (MSK, Nat Genet 2016)\n",
      "Study ID: bcc_unige_2016, Name: Basal Cell Carcinoma (UNIGE, Nat Genet 2016)\n",
      "Study ID: brca_broad, Name: Breast Invasive Carcinoma (Broad, Nature 2012)\n",
      "Study ID: blca_mskcc_solit_2014, Name: Bladder Cancer (MSK, Eur Urol 2014)\n",
      "Study ID: blca_nmibc_2017, Name: Nonmuscle Invasive Bladder Cancer (MSK Eur Urol 2017)\n",
      "Study ID: bfn_duke_nus_2015, Name: Breast Fibroepithelial Tumors (Duke-NUS, Nat Genet 2015)\n",
      "Study ID: brca_jup_msk_2020, Name: Juvenile Papillomatosis and Breast Cancer (MSK,J Pathol. 2015)\n",
      "Study ID: brca_mapk_hp_msk_2021, Name: MAPK on resistance to anti-HER2 therapy for breast cancer (MSK, Nat Commun. 2022)\n",
      "Study ID: brca_hta9_htan_2022, Name: Breast Cancer (HTAN, 2022)\n",
      "Study ID: biliary_tract_summit_2022, Name: Metastatic Biliary Tract Cancers (SUMMIT - Neratinib Basket Trial, 2022)\n",
      "Study ID: bowel_colitis_msk_2022, Name: Colorectal Adenocarcinoma (MSK, Nat Commun 2022)\n",
      "Study ID: bladder_columbia_msk_2018, Name: Bladder Cancer (Columbia University/MSK, Cell 2018)\n",
      "Study ID: bladder_msk_2023, Name: Bladder Cancer (MSK, Clin Cancer Res 2023)\n",
      "Study ID: bm_nsclc_mskcc_2023, Name: Non-Small Cell Lung Cancer Brain Metastasis (MSK, Nat Commun 2023)\n",
      "Study ID: breast_msk_2018, Name: Breast Cancer (MSK, Cancer Cell 2018)\n",
      "Study ID: brca_mskcc_2019, Name: Breast Cancer (MSK, NPJ Breast Cancer 2019)\n",
      "Study ID: breast_alpelisib_2020, Name: Breast Cancer (MSK, Nature Cancer 2020)\n",
      "Study ID: cfdna_msk_2019, Name: Mixed cfDNA (MSK, Nature Medicine 2019)\n",
      "Study ID: ccrcc_dfci_2019, Name: Clear Cell Renal Cell Carcinoma (DFCI, Science 2019)\n",
      "Study ID: breast_ink4_msk_2021, Name: Metastatic Breast Cancer (MSK, Cancer Discovery 2022)\n",
      "Study ID: brca_pareja_msk_2020, Name: Breast Cancer (MSK, Clinical Cancer Res 2020)\n",
      "Study ID: cervix_msk_2023, Name: Cervical Cancer (MSK, Clin Cancer Res 2023)\n",
      "Study ID: chol_jhu_2013, Name: Intrahepatic Cholangiocarcinoma (JHU, Nat Genet 2013)\n",
      "Study ID: chol_nccs_2013, Name: Cholangiocarcinoma (National Cancer Centre of Singapore, Nat Genet 2013)\n",
      "Study ID: chol_nus_2012, Name: Cholangiocarcinoma (National University of Singapore, Nat Genet 2012)\n",
      "Study ID: coadread_mskcc, Name: Colorectal Adenocarcinoma Triplets (MSK, Genome Biol 2014)\n",
      "Study ID: cllsll_icgc_2011, Name: Chronic lymphocytic leukemia (ICGC, Nature Genetics 2011)\n",
      "Study ID: coad_caseccc_2015, Name: Colon Adenocarcinoma (CaseCCC, PNAS 2015)\n",
      "Study ID: chol_msk_2018, Name: Cholangiocarcinoma (MSK, Clin Cancer Res 2018)\n",
      "Study ID: chol_icgc_2017, Name: Cholangiocarcinoma (ICGC, Cancer Discov 2017)\n",
      "Study ID: coadread_mskresistance_2022, Name: Colorectal Cancer (MSK, Cancer Discovery 2022)\n",
      "Study ID: cscc_dfarber_2015, Name: Cutaneous Squamous Cell Carcinoma (DFCI, Clin Cancer Res 2015)\n",
      "Study ID: ctcl_columbia_2015, Name: Cutaneous T Cell Lymphoma (Columbia U, Nat Genet 2015)\n",
      "Study ID: crc_msk_2017, Name: Metastatic Colorectal Cancer (MSK, Cancer Cell 2018)\n",
      "Study ID: crc_eo_2020, Name: Colorectal Cancer (MSK, JNCI 2021)\n",
      "Study ID: crc_apc_impact_2020, Name: Colorectal Cancer (MSK, Gastroenterology 2020)\n",
      "Study ID: crc_nigerian_2020, Name: Disparities in metastatic colorectal cancer between Africans and Americans (MSK, 2020)\n",
      "Study ID: crc_dd_2022, Name: Colorectal Cancer (MSK, JCO Precis Oncol 2022)\n",
      "Study ID: difg_msk_2023, Name: IDH-mutated Diffuse Glioma (MSK, Clin Cancer Res 2024)\n",
      "Study ID: escc_ucla_2014, Name: Esophageal Squamous Cell Carcinoma (UCLA, Nat Genet 2014)\n",
      "Study ID: esca_broad, Name: Esophageal Adenocarcinoma (DFCI, Nat Genet 2013)\n",
      "Study ID: gct_msk_2016, Name: Germ Cell Tumors (MSK, J Clin Oncol 2016)\n",
      "Study ID: egc_msk_2017, Name: Metastatic Esophagogastric Cancer (MSK, Cancer Discovery 2017)\n",
      "Study ID: hcc_mskimpact_2018, Name: Hepatocellular Carcinoma (MSK, Clin Cancer Res 2018)\n",
      "Study ID: hcc_msk_venturaa_2018, Name: Liver Hepatocellular Adenoma and Carcinomas (MSK, PLOS One 2018)\n",
      "Study ID: dlbcl_duke_2017, Name: Diffuse Large B-Cell Lymphoma (Duke, Cell 2017)\n",
      "Study ID: glioma_mskcc_2019, Name: Glioma (MSK, Clin Cancer Res 2019)\n",
      "Study ID: glioma_msk_2018, Name: Glioma (MSK, Nature 2019)\n",
      "Study ID: gbc_msk_2018, Name: Gallbladder Cancer (MSK, Cancer 2018)\n",
      "Study ID: gbm_columbia_2019, Name: Glioblastoma (Columbia, Nat Med. 2019)\n",
      "Study ID: gct_msk_2020, Name: Germ Cell Tumors and Shared Leukemias (MSK, J Clin Invest 2020)\n",
      "Study ID: egc_mskcc_2020, Name: Esophageal/Stomach Cancer (MSK, 2020)\n",
      "Study ID: egc_msk_tp53_ccr_2022, Name: Esophagogastric Cancer (MSK, Clin Cancer Res 2022)\n",
      "Study ID: gbc_mskcc_2022, Name: Gallbladder Cancer (MSK, 2022)\n",
      "Study ID: gist_msk_2022, Name: Gastrointestinal Stromal Tumor (MSK, NPJ Precis Oncol 2023)\n",
      "Study ID: egc_msk_2023, Name: Esophagogastric Cancer (MSK, J Natl Cancer Inst 2023)\n",
      "Study ID: hcc_jcopo_msk_2023, Name: Hepatocellular Carcinoma (MSK, JCO Precis Oncol 2023)\n",
      "Study ID: es_dsrct_msk_2023, Name: Ewing Sarcoma (MSK, 2023)\n",
      "Study ID: kirc_bgi, Name: Kidney Renal Clear Cell Carcinoma (BGI, Nat Genet 2012)\n",
      "Study ID: hnsc_jhu, Name: Head and Neck Squamous Cell Carcinoma (Johns Hopkins, Science 2011)\n",
      "Study ID: hnsc_tcga_pub, Name: Head and Neck Squamous Cell Carcinoma (TCGA, Nature 2015)\n",
      "Study ID: hnc_mskcc_2016, Name: Recurrent and Metastatic Head & Neck Cancer (MSK, JAMA Oncol 2016)\n",
      "Study ID: ihch_smmu_2014, Name: Intrahepatic Cholangiocarcinoma (Shanghai, Nat Commun 2014)\n",
      "Study ID: histiocytosis_cobi_msk_2019, Name: Histiocytosis Cobimetinib (MSK, Nature 2019)\n",
      "Study ID: ihch_mskcc_2020, Name: Intrahepatic Cholangiocarcinoma (MSK, Clin Cancer Res 2021)\n",
      "Study ID: ihch_ismms_2015, Name: Intrahepatic Cholangiocarcinoma (Mount Sinai 2015)\n",
      "Study ID: ihch_msk_2021, Name: Intrahepatic Cholangiocarcinoma (MSK, Hepatology 2021)\n",
      "Study ID: hgsoc_msk_2021, Name: High-Grade Serous Ovarian Cancer (MSK, NPJ Genome Med 2021)\n",
      "Study ID: ilc_msk_2023, Name: Non-CDH1 Invasive Lobular Carcinoma (MSK, 2023)\n",
      "Study ID: lihc_riken, Name: Liver Hepatocellular Carcinoma (RIKEN, Nat Genet 2012)\n",
      "Study ID: liad_inserm_fr_2014, Name: Hepatocellular Adenoma (INSERM, Cancer Cell 2014)\n",
      "Study ID: lcll_broad_2013, Name: Chronic Lymphocytic Leukemia (Broad, Cell 2013)\n",
      "Study ID: lgsoc_mapk_msk_2022, Name: Low-Grade Serous Ovarian Cancer (MSK, Clin Cancer Res 2022)\n",
      "Study ID: luad_tsp, Name: Lung Adenocarcinoma (TSP, Nature 2008)\n",
      "Study ID: lung_msk_2017, Name: Non-Small Cell Cancer (MSK, Cancer Discov 2017)\n",
      "Study ID: lung_msk_pdx, Name: Thoracic PDX (MSK, Provisional)\n",
      "Study ID: lymphoma_cellline_msk_2020, Name: Lymphoma Cell Lines (MSK, Blood 2020)\n",
      "Study ID: lung_msk_mind_2020, Name: Lung Adenocarcinoma (MSK Mind,Nature Cancer 2022)\n",
      "Study ID: mbc_msk_2021, Name: Metaplastic Breast Cancer (MSK,NPJ Breast Cancer 2021)\n",
      "Study ID: luad_mskimpact_2021, Name: Lung Adenocarcinoma (MSK, 2021)\n",
      "Study ID: lung_pdx_msk_2021, Name: Thoracic Cancer (MSK, Nat Commun 2021)\n",
      "Study ID: lung_nci_2022, Name: Lung Cancer in Never Smokers (NCI, Nature Genetics 2021)\n",
      "Study ID: mbl_broad_2012, Name: Medulloblastoma (Broad, Nature 2012)\n",
      "Study ID: mbl_icgc, Name: Medulloblastoma (ICGC, Nature 2012)\n",
      "Study ID: mbl_pcgp, Name: Medulloblastoma (PCGP, Nature 2012)\n",
      "Study ID: mcl_idibips_2013, Name: Mantle Cell Lymphoma (IDIBIPS, PNAS 2013)\n",
      "Study ID: mds_tokyo_2011, Name: Myelodysplasia (UTokyo, Nature 2011)\n",
      "Study ID: mbl_dkfz_2017, Name: Medulloblastoma (DKFZ, Nature 2017)\n",
      "Study ID: mds_iwg_2022, Name: Myelodysplastic Syndromes (MDS IWG, IPSSM, NEJM Evidence 2022)\n",
      "Study ID: alal_target_gdc, Name: Acute Leukemias of Ambiguous Lineage - Phase III (TARGET, GDC)\n",
      "Study ID: aml_target_gdc, Name: Acute Myeloid Leukemia (TARGET, GDC)\n",
      "Study ID: bll_target_gdc, Name: B-Lymphoblastic Leukemia/Lymphoma - Phase II (TARGET, GDC)\n",
      "Study ID: nbl_target_gdc, Name: Neuroblastoma (TARGET, GDC)\n",
      "Study ID: os_target_gdc, Name: Osteosarcoma (TARGET, GDC)\n",
      "Study ID: wt_target_gdc, Name: Wilms' Tumor (TARGET, GDC)\n",
      "Study ID: mpn_cimr_2013, Name: Myeloproliferative Neoplasms (CIMR, NEJM 2013)\n",
      "Study ID: mnm_washu_2016, Name: Acute myeloid leukemia or myelodysplastic syndromes (WashU, 2016)\n",
      "Study ID: metastatic_solid_tumors_mich_2017, Name: Metastatic Solid Cancers (UMich, Nature 2017)\n",
      "Study ID: mixed_selpercatinib_2020, Name: Mixed Tumors Selpercatinib RET Trial (MSK, Nat Commun. 2022)\n",
      "Study ID: mixed_cfdna_msk_2020, Name: Mixed cfDNA (MSK, Genome Med 2021)\n",
      "Study ID: mel_mskimpact_2020, Name: Melanoma (MSK, Clin Cancer Res 2021)\n",
      "Study ID: mixed_kunga_msk_2022, Name: Pediatric Pancan Tumors (MSK,Nat Commun. 2022)\n",
      "Study ID: mixed_impact_subset_2022, Name: Pan-Cancer MSK-IMPACT MET Validation Cohort (MSK 2022)\n",
      "Study ID: nbl_amc_2012, Name: Neuroblastoma (AMC Amsterdam, Nature 2012)\n",
      "Study ID: msk_ch_2020, Name: Cancer Therapy and Clonal Hematopoiesis (MSK, Nat Genet 2020)\n",
      "Study ID: msk_access_2021, Name: MSK-IMPACT and MSK-ACCESS Mixed Cohort (MSK, Nat Commun 2021)\n",
      "Study ID: msk_ch_ped_2021, Name: Cancer Therapy and Clonal Hematopoiesis (MSK, Clin Cancer Res 2022)\n",
      "Study ID: msk_spectrum_tme_2022, Name: Ovarian Cancer - MSK SPECTRUM (MSK, Nature 2022)\n",
      "Study ID: mtnn_msk_2022, Name: Mature T and NK Neoplasms (MSK, Blood Adv 2023)\n",
      "Study ID: msk_ch_2023, Name: Cancer Therapy and Clonal Hematopoiesis (MSK, Blood Adv 2023)\n",
      "Study ID: npc_nusingapore, Name: Nasopharyngeal Carcinoma (Singapore, Nat Genet 2014)\n",
      "Study ID: odg_msk_2017, Name: Anaplastic Oligodendroglioma and Anaplastic Oligoastrocytoma (MSK, Neuro Oncol 2017)\n",
      "Study ID: nsclc_unito_2016, Name: Non-Small Cell Lung Cancer (University of Turin, Lung Cancer 2017)\n",
      "Study ID: nsclc_pd1_msk_2018, Name: Non-Small Cell Lung Cancer (MSK, J Clin Oncol 2018)\n",
      "Study ID: nsclc_mskcc_2015, Name: Non-Small Cell Lung Cancer (MSK, Science 2015)\n",
      "Study ID: nsclc_ctdx_msk_2022, Name: Metastatic Non-Small Cell Lung Cancer (MSK, Nature Medicine 2022)\n",
      "Study ID: paac_msk_jco_2023, Name: Pancreatic Acinar Cell Carcinoma (MSK, JCO 2023)\n",
      "Study ID: panet_jhu_2011, Name: Pancreatic Neuroendocrine Tumors (Johns Hopkins University, Science 2011)\n",
      "Study ID: pcnsl_mayo_2015, Name: Primary Central Nervous System Lymphoma (Mayo Clinic, Clin Cancer Res 2015)\n",
      "Study ID: panet_shanghai_2013, Name: Insulinoma (Shanghai, Nat Commun 2013)\n",
      "Study ID: plmeso_nyu_2015, Name: Pleural Mesothelioma (NYU, Cancer Res 2015)\n",
      "Study ID: past_dkfz_heidelberg_2013, Name: Pilocytic Astrocytoma (ICGC, Nature Genetics 2013)\n",
      "Study ID: pediatric_dkfz_2017, Name: Pediatric Pan-Cancer (DKFZ, Nature 2017)\n",
      "Study ID: paired_bladder_2022, Name: Bladder Cancer (MSK, Cell Reports 2022)\n",
      "Study ID: panet_msk_erc_2023, Name: Pancreatic Neuroendocrine Tumors (MSK, 2023)\n",
      "Study ID: prad_cpcg_2017, Name: Prostate Adenocarcinoma (CPC-GENE, Nature 2017)\n",
      "Study ID: prad_mskcc_2017, Name: Prostate Cancer (MSK, JCO Precis Oncol 2017)\n",
      "Study ID: prad_msk_2019, Name: Prostate Cancer (MSK, Cell Metab 2020)\n",
      "Study ID: prad_mcspc_mskcc_2020, Name: Metastatic castration-sensitive prostate cancer (MSK, Clin Cancer Res 2020)\n",
      "Study ID: scco_mskcc, Name: Small Cell Carcinoma of the Ovary (MSK, Nat Genet 2014)\n",
      "Study ID: rms_nih_2014, Name: Rhabdomyosarcoma (NIH, Cancer Discov 2014)\n",
      "Study ID: sarc_mskcc, Name: Sarcoma (MSKCC/Broad, Nat Genet 2010)\n",
      "Study ID: rectal_msk_2019, Name: Rectal Cancer (MSK,Nature Medicine 2019)\n",
      "Study ID: sarcoma_mskcc_2022, Name: Sarcoma (MSK, Nat Commun. 2022)\n",
      "Study ID: rbl_cfdna_msk_2020, Name: Retinoblastoma cfDNA (MSK, Cancer Med 2020)\n",
      "Study ID: rbl_mskcc_2020, Name: Retinoblastoma (MSK, Cancers 2021)\n",
      "Study ID: prostate_pcbm_swiss_2019, Name: Prostate Cancer Brain Metastases (Bern, Nat Commun. 2022)\n",
      "Study ID: rms_msk_2023, Name: Pediatric Rhabdomyosarcomas (MSK, JCO Precis Oncol 2023)\n",
      "Study ID: sarcoma_msk_2023, Name: Sarcoma (MSK, J Pathol 2023)\n",
      "Study ID: skcm_tcga, Name: Skin Cutaneous Melanoma (TCGA, Firehose Legacy)\n",
      "Study ID: skcm_yale, Name: Skin Cutaneous Melanoma (Yale, Nat Genet 2012)\n",
      "Study ID: skcm_vanderbilt_mskcc_2015, Name: Metastatic Melanoma (MSK, JCO Precis Oncol 2017)\n",
      "Study ID: soft_tissue_msk_2023, Name: Rhabdomyosarcomas (MSK, NPJ Precis Oncol 2023)\n",
      "Study ID: thyroid_mskcc_2016, Name: Poorly-Differentiated and Anaplastic Thyroid Cancers (MSK, JCI 2016)\n",
      "Study ID: summit_2018, Name: SUMMIT - Neratinib Basket Study (Multi-Institute, Nature 2018)\n",
      "Study ID: ucec_msk_2018, Name: Endometrial Cancer (MSK, 2018)\n",
      "Study ID: uccc_nih_2017, Name: Uterine Clear Cell Carcinoma (NIH, Cancer 2017)\n",
      "Study ID: tmb_mskcc_2018, Name: TMB and Immunotherapy (MSK, Nat Genet 2019)\n",
      "Study ID: ucec_ccr_msk_2022, Name: Endometrial Carcinoma MSI (MSK, Clin Cancer Res 2022)\n",
      "Study ID: ucec_ccr_cfdna_msk_2022, Name: Endometrial Carcinoma cfDNA (MSK, Clin Cancer Res 2022)\n",
      "Study ID: ucec_ancestry_cds_msk_2023, Name: Endometrial Cancer (MSK, Cancer Discovery 2023)\n",
      "Study ID: ucec_msk_2024, Name: Endometrial and Ovarian Cancer (MSK, Nature Medicine 2024)\n",
      "Study ID: um_qimr_2016, Name: Uveal Melanoma (QIMR, Oncotarget 2016)\n",
      "Study ID: urcc_mskcc_2016, Name: Unclassified Renal Cell Carcinoma (MSK, Nature 2016)\n",
      "Study ID: utuc_mskcc_2015, Name: Upper Tract Urothelial Cancer (MSK, Eur Urol 2015)\n",
      "Study ID: utuc_msk_2019, Name: Upper Tract Urothelial Carcinoma (MSK, Nat Commun 2020)\n",
      "Study ID: utuc_pdx_msk_2019, Name: Upper Tract Urothelial Carcinoma PDX (MSK, Nat Commun 2020)\n",
      "Study ID: usarc_msk_2020, Name: Uterine Sarcoma/Mesenchymal (MSK, Clin Cancer Res 2020)\n",
      "Study ID: utuc_igbmc_2021, Name: Upper Tract Urothelial Carcinoma (IGBMC, Genome Biology 2021)\n",
      "Study ID: plmeso_msk_2024, Name: Pleural Mesothelioma (MSK, Clin Cancer Res 2024)\n",
      "Study ID: acc_tcga_gdc, Name: Adrenocortical Carcinoma (TCGA, GDC)\n",
      "Study ID: blca_tcga_gdc, Name: Bladder Urothelial Carcinoma (TCGA, GDC)\n",
      "Study ID: brca_tcga_gdc, Name: Invasive Breast Carcinoma (TCGA, GDC)\n",
      "Study ID: cesc_tcga_gdc, Name: Cervical Squamous Cell Carcinoma (TCGA, GDC)\n",
      "Study ID: chol_tcga_gdc, Name: Cholangiocarcinoma (TCGA, GDC)\n",
      "Study ID: dlbclnos_tcga_gdc, Name: Diffuse Large B-Cell Lymphoma, NOS (TCGA, GDC)\n",
      "Study ID: esca_tcga_gdc, Name: Esophageal Adenocarcinoma (TCGA, GDC)\n",
      "Study ID: gbm_tcga_gdc, Name: Glioblastoma Multiforme (TCGA, GDC)\n",
      "Study ID: hnsc_tcga_gdc, Name: Head and Neck Squamous Cell Carcinoma (TCGA, GDC)\n",
      "Study ID: chrcc_tcga_gdc, Name: Chromophobe Renal Cell Carcinoma (TCGA, GDC)\n",
      "Study ID: ccrcc_tcga_gdc, Name: Renal Clear Cell Carcinoma (TCGA, GDC)\n",
      "Study ID: prcc_tcga_gdc, Name: Papillary Renal Cell Carcinoma (TCGA, GDC)\n",
      "Study ID: aml_tcga_gdc, Name: Acute Myeloid Leukemia (TCGA, GDC)\n",
      "Study ID: difg_tcga_gdc, Name: Diffuse Glioma (TCGA, GDC)\n",
      "Study ID: hcc_tcga_gdc, Name: Hepatocellular Carcinoma (TCGA, GDC)\n",
      "Study ID: luad_tcga_gdc, Name: Lung Adenocarcinoma (TCGA, GDC)\n",
      "Study ID: lusc_tcga_gdc, Name: Lung Squamous Cell Carcinoma (TCGA, GDC)\n",
      "Study ID: plmeso_tcga_gdc, Name: Pleural Mesothelioma (TCGA, GDC)\n",
      "Study ID: hgsoc_tcga_gdc, Name: High-Grade Serous Ovarian Cancer (TCGA, GDC)\n",
      "Study ID: paad_tcga_gdc, Name: Pancreatic Adenocarcinoma (TCGA, GDC)\n",
      "Study ID: mnet_tcga_gdc, Name: Miscellaneous Neuroepithelial Tumor (TCGA, GDC)\n",
      "Study ID: prad_tcga_gdc, Name: Prostate Adenocarcinoma (TCGA, GDC)\n",
      "Study ID: read_tcga_gdc, Name: Rectal Adenocarcinoma (TCGA, GDC)\n",
      "Study ID: soft_tissue_tcga_gdc, Name: Soft Tissue Cancer (TCGA, GDC)\n",
      "Study ID: skcm_tcga_gdc, Name: Cutaneous Melanoma (TCGA, GDC)\n",
      "Study ID: stad_tcga_gdc, Name: Stomach Adenocarcinoma (TCGA, GDC)\n",
      "Study ID: nsgct_tcga_gdc, Name: Non-Seminomatous Germ Cell Tumor (TCGA, GDC)\n",
      "Study ID: thpa_tcga_gdc, Name: Papillary Thyroid Cancer (TCGA, GDC)\n",
      "Study ID: thym_tcga_gdc, Name: Thymoma (TCGA, GDC)\n",
      "Study ID: ucec_tcga_gdc, Name: Endometrial Carcinoma (TCGA, GDC)\n",
      "Study ID: ucs_tcga_gdc, Name: Uterine Carcinosarcoma/Uterine Malignant Mixed Mullerian Tumor (TCGA, GDC)\n",
      "Study ID: um_tcga_gdc, Name: Uveal Melanoma (TCGA, GDC)\n",
      "Study ID: brain_cptac_gdc, Name: CNS/Brain Cancer (CPTAC, GDC)\n",
      "Study ID: breast_cptac_gdc, Name: Breast Cancer (CPTAC, GDC)\n",
      "Study ID: coad_cptac_gdc, Name: Colon Adenocarcinoma (CPTAC, GDC)\n",
      "Study ID: luad_cptac_gdc, Name: Lung Adenocarcinoma (CPTAC, GDC)\n",
      "Study ID: lusc_cptac_gdc, Name: Lung Squamous Cell Carcinoma (CPTAC, GDC)\n",
      "Study ID: ohnca_cptac_gdc, Name: Head and Neck Carcinoma, Other (CPTAC, GDC)\n",
      "Study ID: ovary_cptac_gdc, Name: Ovarian Cancer (CPTAC, GDC)\n",
      "Study ID: pancreas_cptac_gdc, Name: Pancreatic Cancer (CPTAC, GDC)\n",
      "Study ID: rcc_cptac_gdc, Name: Renal Cell Carcinoma (CPTAC, GDC)\n",
      "Study ID: uec_cptac_gdc, Name: Uterine Endometrioid Carcinoma (CPTAC, GDC)\n",
      "Study ID: msk_impact_2017, Name: MSK-IMPACT Clinical Sequencing Cohort (MSK, Nat Med 2017)\n",
      "Study ID: coad_tcga_gdc, Name: Colon Adenocarcinoma (TCGA, GDC)\n",
      "Study ID: pancreas_msk_2024, Name: Pancreatic Cancer (MSK, Cancer Cell 2024)\n",
      "Study ID: crc_orion_2024, Name: Colorectal Adenocarcinoma (DFCI/Orion, 2024)\n",
      "Study ID: lms_msk_2024, Name: Leiomyosarcoma  (MSK, Clin Cancer Res 2024)\n",
      "Study ID: prostate_msk_2024, Name: Prostate Cancer (MSK, Clin Cancer Res 2024)\n",
      "Study ID: ucs_msk_2024, Name: ERBB2-mutant Endometrial cancer (MSK,2024)\n",
      "Study ID: panet_msk_2018, Name: Metastatic Pancreatic Neuroendocrine Tumor (MSK, JCO Precis Oncol 2018)\n",
      "Study ID: kirp_tcga, Name: Kidney Renal Papillary Cell Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: ntrk_msk_2019, Name: Tumors with TRK fusions (MSK, Clin Cancer Res 2020)\n",
      "Study ID: heme_msk_impact_2022, Name: MSK-IMPACT Heme Tumors (MSK, 2022)\n",
      "Study ID: makeanimpact_ccr_2023, Name: MSK Make-an-IMPACT Rare Cancers (MSK, Clin Cancer Res 2023)\n",
      "Study ID: pancan_mimsi_msk_2024, Name: Mixed Tumors - MiMSI Cohort (MSK, 2024)\n",
      "Study ID: acc_tcga, Name: Adrenocortical Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: blca_tcga, Name: Bladder Urothelial Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: ampca_bcm_2016, Name: Ampullary Carcinoma (Baylor College of Medicine, Cell Reports 2016)\n",
      "Study ID: blca_dfarber_mskcc_2014, Name: Bladder Urothelial Carcinoma (DFCI/MSK, Cancer Discov 2014)\n",
      "Study ID: blca_mskcc_solit_2012, Name: Bladder Cancer (MSK, J Clin Onco 2013)\n",
      "Study ID: brca_bccrc_xenograft_2014, Name: Breast Cancer Xenografts (British Columbia, Nature 2015)\n",
      "Study ID: blca_bgi, Name: Bladder Urothelial Carcinoma (BGI, Nat Genet 2013)\n",
      "Study ID: blca_tcga_pub, Name: Bladder Urothelial Carcinoma (TCGA, Nature 2014)\n",
      "Study ID: brca_bccrc, Name: Breast Invasive Carcinoma (British Columbia, Nature 2012)\n",
      "Study ID: brca_igr_2015, Name: Metastatic Breast Cancer (INSERM, PLoS Med 2016)\n",
      "Study ID: acbc_mskcc_2015, Name: Adenoid Cystic Carcinoma of the Breast (MSK, J Pathol. 2015)\n",
      "Study ID: acyc_mskcc_2013, Name: Adenoid Cystic Carcinoma (MSK, Nat Genet 2013)\n",
      "Study ID: angs_project_painter_2018, Name: The Angiosarcoma Project - Count Me In (Nature Medicine, 2020)\n",
      "Study ID: all_phase2_target_2018_pub, Name: Pediatric Acute Lymphoid Leukemia - Phase II (TARGET, 2018)\n",
      "Study ID: aml_target_2018_pub, Name: Pediatric Acute Myeloid Leukemia (TARGET, 2018)\n",
      "Study ID: blca_tcga_pub_2017, Name: Bladder Cancer (TCGA, Cell 2017)\n",
      "Study ID: blca_tcga_pan_can_atlas_2018, Name: Bladder Urothelial Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: blca_cornell_2016, Name: Urothelial Carcinoma (Cornell/Trento, Nat Gen 2016)\n",
      "Study ID: aml_ohsu_2018, Name: Acute Myeloid Leukemia (OHSU, Nature 2018)\n",
      "Study ID: acc_2019, Name: Adenoid Cystic Carcinoma Project (J Clin Invest 2019)\n",
      "Study ID: blca_bcan_hcrn_2022, Name: Urothelial Carcinoma (BCAN/HCRN, Nat Commun 2022)\n",
      "Study ID: angs_painter_2020, Name: The Angiosarcoma Project (Provisional, July 2020)\n",
      "Study ID: blca_msk_tcga_2020, Name: Bladder Cancer (MSK/TCGA, 2020)\n",
      "Study ID: brain_cptac_2020, Name: Pediatric Brain Cancer (CPTAC/CHOP, Cell 2020)\n",
      "Study ID: brca_cptac_2020, Name: Proteogenomic landscape of breast cancer (CPTAC, Cell 2020)\n",
      "Study ID: brca_mbcproject_2022, Name: The Metastatic Breast Cancer Project (Provisional, December 2021)\n",
      "Study ID: aml_ohsu_2022, Name: Acute Myeloid Leukemia (OHSU, Cancer Cell 2022)\n",
      "Study ID: asclc_msk_2024, Name: Atypical Small Cell Lung Cancer (MSK, Cancer Discov 2024)\n",
      "Study ID: pcnsl_msk_2024, Name: Primary CNS Lymphoma (MSK, Clin Cancer Res 2024)\n",
      "Study ID: msk_ctdna_vte_2024, Name: MSK ctDNA Sequencing Cohort (MSK, Nature Medicine 2024)\n",
      "Study ID: brca_dfci_2020, Name: Metastatic Breast Cancer (DFCI, Cancer Discov 2020)\n",
      "Study ID: brca_tcga, Name: Breast Invasive Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: cesc_tcga, Name: Cervical Squamous Cell Carcinoma and Endocervical Adenocarcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: chol_tcga, Name: Cholangiocarcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: brca_sanger, Name: Breast Invasive Carcinoma (Sanger, Nature 2012)\n",
      "Study ID: brca_tcga_pub2015, Name: Breast Invasive Carcinoma (TCGA, Cell 2015)\n",
      "Study ID: brca_tcga_pub, Name: Breast Invasive Carcinoma (TCGA, Nature 2012)\n",
      "Study ID: cellline_ccle_broad, Name: Cancer Cell Line Encyclopedia (Novartis/Broad, Nature 2012)\n",
      "Study ID: ccrcc_irc_2014, Name: Kidney Renal Clear Cell Carcinoma (IRC, Nat Genet 2014)\n",
      "Study ID: ccrcc_utokyo_2013, Name: Renal Clear Cell Carcinoma (UTokyo, Nat Genet 2013)\n",
      "Study ID: coadread_genentech, Name: Colorectal Adenocarcinoma (Genentech, Nature 2012)\n",
      "Study ID: cellline_nci60, Name: NCI-60 Cell Lines (NCI, Cancer Res 2012)\n",
      "Study ID: cll_iuopa_2015, Name: Chronic Lymphocytic Leukemia (IUOPA, Nature 2015)\n",
      "Study ID: brca_metabric, Name: Breast Cancer (METABRIC, Nature 2012 & Nat Commun 2016)\n",
      "Study ID: coadread_dfci_2016, Name: Colorectal Adenocarcinoma (DFCI, Cell Reports 2016)\n",
      "Study ID: cll_broad_2015, Name: Chronic Lymphocytic Leukemia (Broad, Nature 2015)\n",
      "Study ID: brca_mbcproject_wagle_2017, Name: The Metastatic Breast Cancer Project (Archived, 2020)\n",
      "Study ID: brca_tcga_pan_can_atlas_2018, Name: Breast Invasive Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: cesc_tcga_pan_can_atlas_2018, Name: Cervical Squamous Cell Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: chol_tcga_pan_can_atlas_2018, Name: Cholangiocarcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: ccle_broad_2019, Name: Cancer Cell Line Encyclopedia (Broad, 2019)\n",
      "Study ID: coad_cptac_2019, Name: Colon Cancer (CPTAC-2 Prospective, Cell 2019)\n",
      "Study ID: brca_smc_2018, Name: Breast Cancer (SMC 2018)\n",
      "Study ID: coadread_cass_2020, Name: Colorectal Cancer (CAS Shanghai, Cancer Cell 2020)\n",
      "Study ID: cll_broad_2022, Name: Chronic Lymphocytic Leukemia (Broad, Nature Genetics 2022)\n",
      "Study ID: coadread_tcga, Name: Colorectal Adenocarcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: dlbc_tcga, Name: Lymphoid Neoplasm Diffuse Large B-cell Lymphoma (TCGA, Firehose Legacy)\n",
      "Study ID: coadread_tcga_pub, Name: Colorectal Adenocarcinoma (TCGA, Nature 2012)\n",
      "Study ID: desm_broad_2015, Name: Desmoplastic Melanoma (Broad Institute, Nat Genet 2015)\n",
      "Study ID: dlbc_broad_2012, Name: Diffuse Large B-Cell Lymphoma (Broad, PNAS 2012)\n",
      "Study ID: cscc_hgsc_bcm_2014, Name: Cutaneous Squamous Cell Carcinoma (MD Anderson, Clin Cancer Res 2014)\n",
      "Study ID: coadread_tcga_pan_can_atlas_2018, Name: Colorectal Adenocarcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: dlbc_tcga_pan_can_atlas_2018, Name: Diffuse Large B-Cell Lymphoma (TCGA, PanCancer Atlas)\n",
      "Study ID: dlbcl_dfci_2018, Name: Diffuse Large B cell Lymphoma (DFCI, Nat Med 2018)\n",
      "Study ID: difg_glass_2019, Name: Diffuse Glioma (GLASS Consortium, Nature 2019)\n",
      "Study ID: cscc_ucsf_2021, Name: Cutaneous Squamous Cell Carcinoma (UCSF, NPJ Genom Med 2021)\n",
      "Study ID: crc_hta11_htan_2021, Name: Pre-cancer Colorectal Polyps (HTAN Vanderbilt, Cell 2021)\n",
      "Study ID: cscc_ranson_2022, Name: Cutaneous Squamous Cell Carcinoma (UOW, Front Oncol 2022)\n",
      "Study ID: difg_glass, Name: Diffuse Glioma (GLASS Consortium)\n",
      "Study ID: esca_tcga, Name: Esophageal Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: escc_icgc, Name: Esophageal Squamous Cell Carcinoma (ICGC, Nature 2014)\n",
      "Study ID: es_dfarber_broad_2014, Name: Pediatric Ewing Sarcoma (DFCI, Cancer Discov 2014)\n",
      "Study ID: es_iocurie_2014, Name: Ewing Sarcoma (Institut Curie, Cancer Discov 2014)\n",
      "Study ID: egc_tmucih_2015, Name: Gastric Adenocarcinoma (TMUCIH, PNAS 2015)\n",
      "Study ID: esca_tcga_pan_can_atlas_2018, Name: Esophageal Adenocarcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: egc_trap_msk_2020, Name: Esophageal Cancer-TRAP Project (MSK, Lancet Oncol 2020)\n",
      "Study ID: egc_trap_ccr_msk_2023, Name: Esophagogastric Cancer (MSK, Clin Cancer Res 2023)\n",
      "Study ID: gbm_tcga, Name: Glioblastoma Multiforme (TCGA, Firehose Legacy)\n",
      "Study ID: gbc_shanghai_2014, Name: Gallbladder Carcinoma (Shanghai, Nat Genet 2014)\n",
      "Study ID: gbm_tcga_pub, Name: Glioblastoma (TCGA, Nature 2008)\n",
      "Study ID: gbm_tcga_pub2013, Name: Glioblastoma (TCGA, Cell 2013)\n",
      "Study ID: gbm_tcga_pan_can_atlas_2018, Name: Glioblastoma Multiforme (TCGA, PanCancer Atlas)\n",
      "Study ID: gbm_mayo_pdx_sarkaria_2019, Name: Brain Tumor PDXs (Mayo Clinic, Clin Cancer Res 2020)\n",
      "Study ID: gbm_cptac_2021, Name: Glioblastoma (CPTAC, Cell 2021)\n",
      "Study ID: gist_msk_2023, Name: Gastrointestinal Stromal Tumors (MSK, Clin Cancer Res 2023)\n",
      "Study ID: kirc_tcga, Name: Kidney Renal Clear Cell Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: kich_tcga, Name: Kidney Chromophobe (TCGA, Firehose Legacy)\n",
      "Study ID: hnsc_tcga, Name: Head and Neck Squamous Cell Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: kirc_tcga_pub, Name: Kidney Renal Clear Cell Carcinoma (TCGA, Nature 2013)\n",
      "Study ID: kich_tcga_pub, Name: Kidney Chromophobe (TCGA, Cancer Cell 2014)\n",
      "Study ID: hnsc_broad, Name: Head and Neck Squamous Cell Carcinoma (Broad, Science 2011)\n",
      "Study ID: hnsc_mdanderson_2013, Name: Oral Squamous Cell Carcinoma (MD Anderson, Cancer Discov 2013)\n",
      "Study ID: hcc_inserm_fr_2015, Name: Hepatocellular Carcinomas (INSERM, Nat Genet 2015)\n",
      "Study ID: hccihch_pku_2019, Name: Combined Hepatocellular and Intrahepatic Cholangiocarcinoma (Peking University, Cancer Cell 2019)\n",
      "Study ID: hcc_meric_2021, Name: Hepatocellular Carcinoma (MERiC/Basel, Nat Commun. 2022)\n",
      "Study ID: hcc_clca_2024, Name: Hepatocellular Carcinoma (CLCA, Nature 2024)\n",
      "Study ID: hcc_msk_2024, Name: Hepatocellular Carcinoma (MSK, Clin Cancer Res 2024)\n",
      "Study ID: laml_tcga, Name: Acute Myeloid Leukemia (TCGA, Firehose Legacy)\n",
      "Study ID: lgg_tcga, Name: Brain Lower Grade Glioma (TCGA, Firehose Legacy)\n",
      "Study ID: lihc_tcga, Name: Liver Hepatocellular Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: luad_tcga, Name: Lung Adenocarcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: laml_tcga_pub, Name: Acute Myeloid Leukemia (TCGA, NEJM 2013)\n",
      "Study ID: lgg_ucsf_2014, Name: Low-Grade Gliomas (UCSF, Science 2014)\n",
      "Study ID: lgggbm_tcga_pub, Name: Merged Cohort of LGG and GBM (TCGA, Cell 2016)\n",
      "Study ID: lihc_amc_prv, Name: Liver Hepatocellular Carcinoma (AMC, Hepatology 2014)\n",
      "Study ID: luad_mskcc_2015, Name: Lung Adenocarcinoma (MSK, Science 2015)\n",
      "Study ID: luad_broad, Name: Lung Adenocarcinoma (Broad, Cell 2012)\n",
      "Study ID: luad_mskcc_2020, Name: Lung Adenocarcinoma (MSK, J Thorac Oncol 2020)\n",
      "Study ID: luad_mskcc_2023_met_organotropism, Name: Lung Adenocarcinoma Met Organotropism (MSK, Cancer Cell 2023)\n",
      "Study ID: luad_oncosg_2020, Name: Lung Adenocarcinoma (OncoSG, Nat Genet 2020)\n",
      "Study ID: luad_msk_npjpo_2021, Name: Lung Adenocarcinoma (MSK, NPJ Precision Oncology 2021)\n",
      "Study ID: luad_cptac_2020, Name: Lung Adenocarcinoma (CPTAC, Cell 2020)\n",
      "Study ID: lusc_tcga, Name: Lung Squamous Cell Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: meso_tcga, Name: Mesothelioma (TCGA, Firehose Legacy)\n",
      "Study ID: luad_tcga_pub, Name: Lung Adenocarcinoma (TCGA, Nature 2014)\n",
      "Study ID: lusc_tcga_pub, Name: Lung Squamous Cell Carcinoma (TCGA, Nature 2012)\n",
      "Study ID: mm_broad, Name: Multiple Myeloma (Broad, Cancer Cell 2014)\n",
      "Study ID: mpnst_mskcc, Name: Malignant Peripheral Nerve Sheath Tumor (MSK, Nat Genet 2014)\n",
      "Study ID: mbl_sickkids_2016, Name: Medulloblastoma (Sickkids, Nature 2016)\n",
      "Study ID: mrt_bcgsc_2016, Name: Rhabdoid Cancer (BCGSC, Cancer Cell 2016)\n",
      "Study ID: mel_tsam_liang_2017, Name: Acral Melanoma (TGEN, Genome Res 2017)\n",
      "Study ID: mel_ucla_2016, Name: Metastatic Melanoma (UCLA, Cell 2016)\n",
      "Study ID: mixed_pipseq_2017, Name: Pediatric Pan-cancer (Columbia U, Genome Med 2016)\n",
      "Study ID: mixed_allen_2018, Name: MSS Mixed Solid Tumors (Broad/Dana-Farber, Nat Genet 2018)\n",
      "Study ID: mbn_mdacc_2013, Name: Mature B-cell malignancies (MD Anderson Cancer Center)\n",
      "Study ID: mds_mskcc_2020, Name: Myelodysplastic (MSK, 2020)\n",
      "Study ID: mel_dfci_2019, Name: Metastatic Melanoma (DFCI, Nature Medicine 2019)\n",
      "Study ID: lung_smc_2016, Name: Lung Cancer (SMC, Cancer Research 2016)\n",
      "Study ID: mixed_msk_tcga_2021, Name: RAD51B Associated Mixed Cancers (MSK, NPJ Breast Cancer 2021)\n",
      "Study ID: mng_utoronto_2021, Name: Meningioma (University of Toronto, Nature 2021)\n",
      "Study ID: mpcproject_broad_2021, Name: The Metastatic Prostate Cancer Project (Provisional, June 2021)\n",
      "Study ID: lusc_cptac_2021, Name: Lung Squamous Cell Carcinoma (CPTAC, Cell 2021)\n",
      "Study ID: mbn_sfu_2023, Name: Mature B-Cell Neoplasms (Simon Fraser University, Blood 2023)\n",
      "Study ID: mbn_msk_2024, Name: Diffuse Large B-Cell Lymphoma (MSK, Haematologica 2024)\n",
      "Study ID: ov_tcga, Name: Ovarian Serous Cystadenocarcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: paad_tcga, Name: Pancreatic Adenocarcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: nccrcc_genentech_2014, Name: Renal Non-Clear Cell Carcinoma (Genentech, Nat Genet 2014)\n",
      "Study ID: ov_tcga_pub, Name: Ovarian Serous Cystadenocarcinoma (TCGA, Nature 2011)\n",
      "Study ID: paac_jhu_2014, Name: Acinar Cell Carcinoma of the Pancreas (JHU, J Pathol 2014)\n",
      "Study ID: paad_icgc, Name: Pancreatic Adenocarcinoma (ICGC, Nature 2012)\n",
      "Study ID: paad_utsw_2015, Name: Pancreatic Cancer (UTSW, Nat Commun 2015)\n",
      "Study ID: nepc_wcm_2016, Name: Neuroendocrine Prostate Cancer (Multi-Institute, Nat Med 2016)\n",
      "Study ID: nbl_ucologne_2015, Name: Neuroblastoma (Broad, Nature 2015)\n",
      "Study ID: nsclc_tcga_broad_2016, Name: Pan-Lung Cancer (TCGA, Nat Genet 2016)\n",
      "Study ID: paad_qcmg_uq_2016, Name: Pancreatic Adenocarcinoma (QCMG, Nature 2016)\n",
      "Study ID: pact_jhu_2011, Name: Cystic Tumor of the Pancreas (Johns Hopkins, PNAS 2011)\n",
      "Study ID: nbl_broad_2013, Name: Neuroblastoma (Broad, Nat Genet 2013)\n",
      "Study ID: nhl_bcgsc_2011, Name: Non-Hodgkin Lymphoma (BCGSC, Nature 2011)\n",
      "Study ID: nhl_bcgsc_2013, Name: Diffuse Large B-cell Lymphoma (BCGSC, Blood 2013)\n",
      "Study ID: nsclc_tracerx_2017, Name: Non-Small Cell Lung Cancer (TRACERx, NEJM & Nature 2017)\n",
      "Study ID: nbl_target_2018_pub, Name: Pediatric Neuroblastoma (TARGET, 2018)\n",
      "Study ID: nsclc_mskcc_2018, Name: Non-Small Cell Lung Cancer (MSK, Cancer Cell 2018)\n",
      "Study ID: paad_cptac_2021, Name: Pancreatic Ductal Adenocarcinoma (CPTAC, Cell 2021)\n",
      "Study ID: pcpg_tcga, Name: Pheochromocytoma and Paraganglioma (TCGA, Firehose Legacy)\n",
      "Study ID: prad_tcga, Name: Prostate Adenocarcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: prad_fhcrc, Name: Prostate Adenocarcinoma (Fred Hutchinson CRC, Nat Med 2016)\n",
      "Study ID: prad_broad, Name: Prostate Adenocarcinoma (Broad/Cornell, Nat Genet 2012)\n",
      "Study ID: prad_broad_2013, Name: Prostate Adenocarcinoma (Broad/Cornell, Cell 2013)\n",
      "Study ID: prad_mich, Name: Metastatic Prostate Adenocarcinoma (MCTP, Nature 2012)\n",
      "Study ID: prad_mskcc, Name: Prostate Adenocarcinoma (MSK, Cancer Cell 2010)\n",
      "Study ID: prad_mskcc_2014, Name: Prostate Adenocarcinoma (MSK, PNAS 2014)\n",
      "Study ID: prad_su2c_2015, Name: Metastatic Prostate Cancer (SU2C/PCF Dream Team, Cell 2015)\n",
      "Study ID: prad_tcga_pub, Name: Prostate Adenocarcinoma (TCGA, Cell 2015)\n",
      "Study ID: panet_arcnet_2017, Name: Pancreatic Neuroendocrine Tumors (Multi-Institute, Nature 2017)\n",
      "Study ID: pcpg_tcga_pub, Name: Pheochromocytoma and Paraganglioma (TCGA, Cell 2017)\n",
      "Study ID: prad_eururol_2017, Name: Prostate Adenocarcinoma (SMMU, Eur Urol 2017)\n",
      "Study ID: prad_p1000, Name: Prostate Adenocarcinoma (MSK/DFCI, Nature Genetics 2018)\n",
      "Study ID: prad_su2c_2019, Name: Metastatic Prostate Adenocarcinoma (SU2C/PCF Dream Team, PNAS 2019)\n",
      "Study ID: prostate_dkfz_2018, Name: Prostate Cancer (DKFZ, Cancer Cell 2018)\n",
      "Study ID: pptc_2019, Name: Pediatric Preclinical Testing Consortium  (CHOP, Cell Rep 2019)\n",
      "Study ID: prad_cdk12_mskcc_2020, Name: Prostate Adenocarcinoma (MSK, Eur Urol 2020)\n",
      "Study ID: prad_mskcc_cheny1_organoids_2014, Name: Prostate Adenocarcinoma Organoids (MSK, Cell 2014)\n",
      "Study ID: pan_origimed_2020, Name: China Pan-cancer (OrigiMed, Nature 2022)\n",
      "Study ID: prad_msk_stopsack_2021, Name: Race Differences in Prostate Cancer (MSK, 2021)\n",
      "Study ID: prad_pik3r1_msk_2021, Name: Prostate Adenocarcinoma (MSK, Clin Cancer Res. 2022)\n",
      "Study ID: pog570_bcgsc_2020, Name: Pan-cancer Analysis of Advanced and Metastatic Tumors (BCGSC, Nature Cancer 2020)\n",
      "Study ID: prad_organoids_msk_2022, Name: Prostate Cancer (MSK, Science 2022)\n",
      "Study ID: ptad_msk_2024, Name: Pituitary Adenoma (MSK, Acta Neuropathologica 2024)\n",
      "Study ID: prad_msk_mdanderson_2023, Name: Prostate Cancer MDA PCa PDX (MD Anderson, Clin Cancer Res 2024)\n",
      "Study ID: sarc_tcga, Name: Sarcoma (TCGA, Firehose Legacy)\n",
      "Study ID: stad_tcga, Name: Stomach Adenocarcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: tgct_tcga, Name: Testicular Germ Cell Cancer (TCGA, Firehose Legacy)\n",
      "Study ID: thym_tcga, Name: Thymoma (TCGA, Firehose Legacy)\n",
      "Study ID: thca_tcga, Name: Thyroid Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: ucec_tcga, Name: Uterine Corpus Endometrial Carcinoma (TCGA, Firehose Legacy)\n",
      "Study ID: ucs_tcga, Name: Uterine Carcinosarcoma (TCGA, Firehose Legacy)\n",
      "Study ID: uvm_tcga, Name: Uveal Melanoma (TCGA, Firehose Legacy)\n",
      "Study ID: sclc_clcgp, Name: Small Cell Lung Cancer (CLCGP, Nat Genet 2012)\n",
      "Study ID: sclc_jhu, Name: Small Cell Lung Cancer (Johns Hopkins, Nat Genet 2012)\n",
      "Study ID: skcm_broad, Name: Skin Cutaneous Melanoma (Broad, Cell 2012)\n",
      "Study ID: skcm_broad_dfarber, Name: Melanoma (Broad/Dana Farber, Nature 2012)\n",
      "Study ID: stad_pfizer_uhongkong, Name: Stomach Adenocarcinoma (Pfizer and UHK, Nat Genet 2014)\n",
      "Study ID: stad_tcga_pub, Name: Stomach Adenocarcinoma (TCGA, Nature 2014)\n",
      "Study ID: stad_uhongkong, Name: Stomach Adenocarcinoma (UHK, Nat Genet 2011)\n",
      "Study ID: stad_utokyo, Name: Stomach Adenocarcinoma (U Tokyo, Nat Genet 2014)\n",
      "Study ID: tet_nci_2014, Name: Thymic Epithelial Tumors (NCI, Nat Genet 2014)\n",
      "Study ID: thca_tcga_pub, Name: Papillary Thyroid Carcinoma (TCGA, Cell 2014)\n",
      "Study ID: ucs_jhu_2014, Name: Uterine Carcinosarcoma (Johns Hopkins, Nat Commun 2014)\n",
      "Study ID: ucec_tcga_pub, Name: Uterine Corpus Endometrial Carcinoma (TCGA, Nature 2013)\n",
      "Study ID: stes_tcga_pub, Name: Esophageal Carcinoma (TCGA, Nature 2017)\n",
      "Study ID: skcm_broad_brafresist_2012, Name: Skin Cutaneous Melanoma(Broad, Cancer Discov 2014)\n",
      "Study ID: sarc_tcga_pub, Name: Adult Soft Tissue Sarcomas (TCGA, Cell 2017)\n",
      "Study ID: skcm_mskcc_2014, Name: Melanoma (MSK, NEJM 2014)\n",
      "Study ID: sclc_cancercell_gardner_2017, Name: Small-Cell Lung Cancer (Multi-Institute, Cancer Cell 2017)\n",
      "Study ID: rt_target_2018_pub, Name: Pediatric Rhabdoid Tumor (TARGET, 2018)\n",
      "Study ID: wt_target_2018_pub, Name: Pediatric Wilms' Tumor (TARGET, 2018)\n",
      "Study ID: skcm_tcga_pub_2015, Name: Melanomas (TCGA, Cell 2015)\n",
      "Study ID: vsc_cuk_2018, Name: Squamous Cell Carcinoma of the Vulva (CUK, Exp Mol Med 2018)\n",
      "Study ID: utuc_cornell_baylor_mdacc_2019, Name: Upper Tract Urothelial Carcinoma (Cornell/Baylor/MDACC, Nat Commun 2019)\n",
      "Study ID: skcm_dfci_2015, Name: Metastatic Melanoma (DFCI, Science 2015)\n",
      "Study ID: sclc_ucologne_2015, Name: Small Cell Lung Cancer (U Cologne, Nature 2015)\n",
      "Study ID: stad_oncosg_2018, Name: Gastric Cancer (OncoSG, 2018)\n",
      "Study ID: rectal_msk_2022, Name: Rectal Cancer (MSK, Nature Medicine 2022)\n",
      "Study ID: ucec_cptac_2020, Name: Endometrial Carcinoma (CPTAC, Cell 2020)\n",
      "Study ID: stmyec_wcm_2022, Name: Myoepithelial Carcinomas of Soft Tissue (WCM, CSH Molecular Case Studies 2022)\n",
      "Study ID: sarcoma_msk_2022, Name: Soft Tissue and Bone Sarcoma (MSK, Nat Commun 2022)\n",
      "Study ID: hnsc_tcga_pan_can_atlas_2018, Name: Head and Neck Squamous Cell Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: kich_tcga_pan_can_atlas_2018, Name: Kidney Chromophobe (TCGA, PanCancer Atlas)\n",
      "Study ID: kirc_tcga_pan_can_atlas_2018, Name: Kidney Renal Clear Cell Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: kirp_tcga_pan_can_atlas_2018, Name: Kidney Renal Papillary Cell Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: laml_tcga_pan_can_atlas_2018, Name: Acute Myeloid Leukemia (TCGA, PanCancer Atlas)\n",
      "Study ID: lgg_tcga_pan_can_atlas_2018, Name: Brain Lower Grade Glioma (TCGA, PanCancer Atlas)\n",
      "Study ID: lihc_tcga_pan_can_atlas_2018, Name: Liver Hepatocellular Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: luad_tcga_pan_can_atlas_2018, Name: Lung Adenocarcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: lusc_tcga_pan_can_atlas_2018, Name: Lung Squamous Cell Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: meso_tcga_pan_can_atlas_2018, Name: Mesothelioma (TCGA, PanCancer Atlas)\n",
      "Study ID: ov_tcga_pan_can_atlas_2018, Name: Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: paad_tcga_pan_can_atlas_2018, Name: Pancreatic Adenocarcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: pcpg_tcga_pan_can_atlas_2018, Name: Pheochromocytoma and Paraganglioma (TCGA, PanCancer Atlas)\n",
      "Study ID: prad_tcga_pan_can_atlas_2018, Name: Prostate Adenocarcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: sarc_tcga_pan_can_atlas_2018, Name: Sarcoma (TCGA, PanCancer Atlas)\n",
      "Study ID: nst_nfosi_ntap, Name: Nerve Sheath Tumors (Johns Hopkins, 2024)\n",
      "Study ID: skcm_tcga_pan_can_atlas_2018, Name: Skin Cutaneous Melanoma (TCGA, PanCancer Atlas)\n",
      "Study ID: stad_tcga_pan_can_atlas_2018, Name: Stomach Adenocarcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: tgct_tcga_pan_can_atlas_2018, Name: Testicular Germ Cell Tumors (TCGA, PanCancer Atlas)\n",
      "Study ID: thca_tcga_pan_can_atlas_2018, Name: Thyroid Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: thym_tcga_pan_can_atlas_2018, Name: Thymoma (TCGA, PanCancer Atlas)\n",
      "Study ID: ucec_tcga_pan_can_atlas_2018, Name: Uterine Corpus Endometrial Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: ucs_tcga_pan_can_atlas_2018, Name: Uterine Carcinosarcoma (TCGA, PanCancer Atlas)\n",
      "Study ID: uvm_tcga_pan_can_atlas_2018, Name: Uveal Melanoma (TCGA, PanCancer Atlas)\n",
      "Study ID: coad_silu_2022, Name: Colon Cancer (Sidra-LUMC AC-ICAM, Nat Med 2023)\n",
      "Study ID: acc_tcga_pan_can_atlas_2018, Name: Adrenocortical Carcinoma (TCGA, PanCancer Atlas)\n",
      "Study ID: msk_chord_2024, Name: MSK-CHORD (MSK, Nature 2024)\n",
      "Study ID: pancan_mappyacts_2022, Name: Pediatric European MAPPYACTS Trial (Gustave Roussy, Cancer Discov 2022)\n",
      "Study ID: msk_met_2021, Name: MSK MetTropism (MSK, Cell 2021)\n",
      "Study ID: rectal_radiation_msk_2024, Name: Colorectal Cancer Radiation (MSK, 2024)\n",
      "Study ID: breast_msk_2025, Name: Breast Cancer (MSK, 2025)\n",
      "Study ID: blca_msk_2024, Name: Bladder Urothelial Carcinoma (MSKCC, JCO Precis Oncol 2024)\n",
      "Study ID: brca_fuscc_2020, Name: Triple-Negative Breast Cancer (FUSCC, Cell Research 2020)\n",
      "Study ID: thyroid_gatci_2024, Name: Anaplastic Thyroid Cancers (GATCI, Cell Reports 2024)\n",
      "Study ID: brca_aurora_2023, Name: Metastatic Breast Cancer (AURORA US Network, Nat Cancer 2023)\n",
      "Study ID: pancan_ped_mai_msk_2025, Name: Pediatric Tumors (MSK, 2025)\n",
      "Study ID: soft_tissue_msk_2025, Name: Soft Tissue Sarcoma (MSK, 2025)\n",
      "Study ID: blca_msk_2025, Name: Bladder Cancer (MSK, Eur Urol Open Sci 2025)\n",
      "Study ID: ucec_msk_2025, Name: Serous Ovarian Cancer (MSK, 2025)\n",
      "Study ID: ovary_geomx_gray_foundation_2024, Name: Ovarian Cancer (Gray Foundation, Cancer Discov 2024)\n",
      "Study ID: pdac_msk_2024, Name: Pancreatic Adenocarcinoma (MSK, Nat Med 2024)\n",
      "Study ID: braf_msk_archer_2024, Name: BRAF Fusions - ARCHER Clinical Sequencing Cohort (MSK, Clin Cancer Res 2024)\n",
      "Study ID: braf_msk_impact_2024, Name: BRAF Fusions - IMPACT Clinical Sequencing Cohort (MSK, Clin Cancer Res 2024)\n",
      "Study ID: sarcoma_ucla_2024, Name: Sarcoma (UCLA, Cell 2024)\n",
      "Study ID: csf_msk_2024, Name: Cerebrospinal Fluid Circulating Tumor DNA (MSK, Acta Neuropathol Commun 2024)\n",
      "Study ID: chl_sccc_2023, Name: Classical Hodgkins Lymphoma (SCCC, Blood Cancer Discov 2023)\n",
      "Study ID: ovarian_msk_2025, Name: Serous Ovarian Cancer (MSK, 2025)\n",
      "Study ID: normal_skin_melanocytes_2024, Name: Normal Melanocytes (UCSF, Nature 2020)\n",
      "Study ID: gist_msk_2025, Name: Gastrointestinal Stromal Tumor (MSK 2025)\n",
      "Study ID: normal_skin_fibroblast_2024, Name: Normal Fibroblasts (UCSF, BioRxiv 2024)\n",
      "Study ID: normal_skin_keratinocytes_2024, Name: Normal Keratinocytes (UCSF, BioRxiv 2024)\n",
      "Study ID: pancan_pcawg_2020, Name: Pan-cancer analysis of whole genomes (ICGC/TCGA, Nature 2020)\n",
      "Study ID: uec_msk_2024, Name: Endometrial cancer HR-positive (MSK, 2024)\n",
      "\n",
      "Loaded 1108 samples\n"
     ]
    }
   ],
   "source": [
    "# Get all studies\n",
    "studies = get_cbioportal_data('studies')\n",
    "print(\"Available studies:\")\n",
    "for study in studies:\n",
    "    print(f\"Study ID: {study['studyId']}, Name: {study['name']}\")\n",
    "\n",
    "# Get samples for BRCA study\n",
    "study_id = 'brca_tcga'\n",
    "samples = get_cbioportal_data(f'studies/{study_id}/samples')\n",
    "sample_ids = [sample['sampleId'] for sample in samples]\n",
    "print(f\"\\nLoaded {len(sample_ids)} samples\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd54967f",
   "metadata": {},
   "source": [
    "## ðŸ§¬ Download RNA-seq Gene Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375ea4ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking available molecular profiles...\n",
      "Available profiles: [{'molecularAlterationType': 'PROTEIN_LEVEL', 'datatype': 'LOG2-VALUE', 'name': 'Protein expression (RPPA)', 'description': 'Protein expression measured by reverse-phase protein array', 'showProfileInAnalysisTab': False, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_rppa', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'PROTEIN_LEVEL', 'datatype': 'Z-SCORE', 'name': 'Protein expression z-scores (RPPA)', 'description': 'Protein expression, measured by reverse-phase protein array, z-scores', 'showProfileInAnalysisTab': True, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_rppa_Zscores', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'PROTEIN_LEVEL', 'datatype': 'CONTINUOUS', 'name': 'Protein levels (mass spectrometry by CPTAC)', 'description': 'Protein levels measured with mass spectrometry by the Clinical Proteomic Tumor Analysis Consortium (CPTAC)', 'showProfileInAnalysisTab': False, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_protein_quantification', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'PROTEIN_LEVEL', 'datatype': 'Z-SCORE', 'name': 'Protein level z-scores (mass spectrometry by CPTAC)', 'description': 'Protein level z-scores measured with mass spectrometry by the Clinical Proteomic Tumor Analysis Consortium (CPTAC)', 'showProfileInAnalysisTab': True, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_protein_quantification_zscores', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'GENERIC_ASSAY', 'genericAssayType': 'PHOSPHOSITE_QUANTIFICATION', 'datatype': 'LIMIT-VALUE', 'name': 'Phosphoprotein site level expression (mass spectrometry by CPTAC)', 'description': 'Phosphoprotein site level expression data measured with mass spectrometry by the Clinical Proteomic Tumor Analysis Consortium (CPTAC)', 'showProfileInAnalysisTab': True, 'pivotThreshold': 0.0, 'sortOrder': 'ASC', 'patientLevel': False, 'molecularProfileId': 'brca_tcga_phosphoprotein_quantification', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'COPY_NUMBER_ALTERATION', 'datatype': 'DISCRETE', 'name': 'Putative copy-number alterations from GISTIC', 'description': 'Putative copy-number calls on 1080 cases determined using GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification.', 'showProfileInAnalysisTab': True, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_gistic', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'COPY_NUMBER_ALTERATION', 'datatype': 'CONTINUOUS', 'name': 'Capped relative linear copy-number values', 'description': 'Capped relative linear copy-number values for each gene (from Affymetrix SNP6).', 'showProfileInAnalysisTab': False, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_linear_CNA', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'MUTATION_EXTENDED', 'datatype': 'MAF', 'name': 'Mutations', 'description': 'Mutation data from whole exome sequencing. Mutation packager: BRCA/20160128/gdac.broadinstitute.org_BRCA.Mutation_Packager_Calls.Level_3.2016012800.0.0.tar.gz.', 'showProfileInAnalysisTab': True, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_mutations', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'METHYLATION', 'datatype': 'CONTINUOUS', 'name': 'Methylation (HM450)', 'description': 'Methylation (HM450) beta-values for genes in 885 cases. For genes with multiple methylation probes, the probe most anti-correlated with expression.', 'showProfileInAnalysisTab': False, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_methylation_hm450', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'MRNA_EXPRESSION', 'datatype': 'CONTINUOUS', 'name': 'mRNA expression (microarray)', 'description': 'Expression levels for 17155 genes in 590 brca cases (Agilent microarray).', 'showProfileInAnalysisTab': False, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_mrna', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'MRNA_EXPRESSION', 'datatype': 'Z-SCORE', 'name': 'mRNA expression z-scores relative to diploid samples (microarray)', 'description': 'mRNA expression z-scores (Agilent microarray) compared to the expression distribution of each gene tumors that are diploid for this gene.', 'showProfileInAnalysisTab': True, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_mrna_median_Zscores', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'MRNA_EXPRESSION', 'datatype': 'Z-SCORE', 'name': 'mRNA expression z-scores relative to all samples (log microarray)', 'description': 'Log-transformed mRNA expression z-scores compared to the expression distribution of all samples  (Agilent microarray).', 'showProfileInAnalysisTab': True, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_mrna_median_all_sample_Zscores', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'MRNA_EXPRESSION', 'datatype': 'CONTINUOUS', 'name': 'mRNA expression (RNA Seq V2 RSEM)', 'description': 'mRNA gene expression (RNA Seq V2 RSEM)', 'showProfileInAnalysisTab': False, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_rna_seq_v2_mrna', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'MRNA_EXPRESSION', 'datatype': 'Z-SCORE', 'name': 'mRNA expression z-scores relative to diploid samples (RNA Seq V2 RSEM)', 'description': 'mRNA expression z-scores (RNA Seq V2 RSEM) compared to the expression distribution of each gene tumors that are diploid for this gene.', 'showProfileInAnalysisTab': True, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_rna_seq_v2_mrna_median_Zscores', 'studyId': 'brca_tcga'}, {'molecularAlterationType': 'MRNA_EXPRESSION', 'datatype': 'Z-SCORE', 'name': 'mRNA expression z-scores relative to all samples (log RNA Seq V2 RSEM)', 'description': 'Log-transformed mRNA expression z-scores compared to the expression distribution of all samples (RNA Seq V2 RSEM).', 'showProfileInAnalysisTab': True, 'patientLevel': False, 'molecularProfileId': 'brca_tcga_rna_seq_v2_mrna_median_all_sample_Zscores', 'studyId': 'brca_tcga'}]\n",
      "\n",
      "Trying with minimal data...\n",
      "Test response status: 200\n",
      "Test response content: \n",
      "\n",
      "Trying with all data...\n",
      "Full response status: 200\n",
      "Full response content: \n"
     ]
    }
   ],
   "source": [
    "genes = ['CD8A', 'PDCD1', 'CTLA4']\n",
    "profile_id = 'brca_tcga_rna_seq_v2_mrna'\n",
    "\n",
    "# Try the fetch endpoint\n",
    "print(\"Trying to get molecular data...\")\n",
    "response = requests.post(\n",
    "    f'{base_url}/molecular-data/fetch',\n",
    "    json={\n",
    "        'molecularProfileId': profile_id,\n",
    "        'sampleIds': sample_ids,\n",
    "        'geneIds': genes\n",
    "    }\n",
    ")\n",
    "print(\"Response status:\", response.status_code)\n",
    "print(\"Response content:\", response.text[:1000])\n",
    "\n",
    "# Try to parse the response\n",
    "molecular_data = response.json()\n",
    "print(\"\\nMolecular data type:\", type(molecular_data))\n",
    "print(\"Molecular data structure:\", molecular_data[:5] if isinstance(molecular_data, list) else molecular_data)\n",
    "\n",
    "# Create DataFrame\n",
    "import pandas as pd\n",
    "if isinstance(molecular_data, list):\n",
    "    expr_df = pd.DataFrame(molecular_data)\n",
    "    print(\"\\nDataFrame columns:\", expr_df.columns)\n",
    "    if all(col in expr_df.columns for col in ['sampleId', 'geneId', 'value']):\n",
    "        expr_df = expr_df.pivot(index='sampleId', columns='geneId', values='value').dropna()\n",
    "        print(\"Final DataFrame shape:\", expr_df.shape)\n",
    "    else:\n",
    "        print(\"Missing required columns. Available columns:\", expr_df.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a9791e8",
   "metadata": {},
   "source": [
    "## ðŸ§¬ Merge Immune Subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d737b3b",
   "metadata": {},
   "outputs": [
    {
     "ename": "ParserError",
     "evalue": "Error tokenizing data. C error: Expected 1 fields in line 18, saw 2\n",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mParserError\u001b[39m                               Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[35]\u001b[39m\u001b[32m, line 14\u001b[39m\n\u001b[32m     11\u001b[39m response = requests.get(url, headers=headers)\n\u001b[32m     12\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m response.status_code == \u001b[32m200\u001b[39m:\n\u001b[32m     13\u001b[39m     \u001b[38;5;66;03m# Read the CSV file\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m14\u001b[39m     subtypes = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mio\u001b[49m\u001b[43m.\u001b[49m\u001b[43mStringIO\u001b[49m\u001b[43m(\u001b[49m\u001b[43mresponse\u001b[49m\u001b[43m.\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     15\u001b[39m     \u001b[38;5;28mprint\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33mSuccessfully downloaded immune subtypes data\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m     16\u001b[39m     \u001b[38;5;28mprint\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33mData shape:\u001b[39m\u001b[33m\"\u001b[39m, subtypes.shape)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.12.4/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1013\u001b[39m kwds_defaults = _refine_defaults_read(\n\u001b[32m   1014\u001b[39m     dialect,\n\u001b[32m   1015\u001b[39m     delimiter,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1022\u001b[39m     dtype_backend=dtype_backend,\n\u001b[32m   1023\u001b[39m )\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.12.4/lib/python3.12/site-packages/pandas/io/parsers/readers.py:626\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    623\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n\u001b[32m    625\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[32m--> \u001b[39m\u001b[32m626\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.12.4/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1923\u001b[39m, in \u001b[36mTextFileReader.read\u001b[39m\u001b[34m(self, nrows)\u001b[39m\n\u001b[32m   1916\u001b[39m nrows = validate_integer(\u001b[33m\"\u001b[39m\u001b[33mnrows\u001b[39m\u001b[33m\"\u001b[39m, nrows)\n\u001b[32m   1917\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m   1918\u001b[39m     \u001b[38;5;66;03m# error: \"ParserBase\" has no attribute \"read\"\u001b[39;00m\n\u001b[32m   1919\u001b[39m     (\n\u001b[32m   1920\u001b[39m         index,\n\u001b[32m   1921\u001b[39m         columns,\n\u001b[32m   1922\u001b[39m         col_dict,\n\u001b[32m-> \u001b[39m\u001b[32m1923\u001b[39m     ) = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[attr-defined]\u001b[39;49;00m\n\u001b[32m   1924\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\n\u001b[32m   1925\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1926\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[32m   1927\u001b[39m     \u001b[38;5;28mself\u001b[39m.close()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.12.4/lib/python3.12/site-packages/pandas/io/parsers/c_parser_wrapper.py:234\u001b[39m, in \u001b[36mCParserWrapper.read\u001b[39m\u001b[34m(self, nrows)\u001b[39m\n\u001b[32m    232\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m    233\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.low_memory:\n\u001b[32m--> \u001b[39m\u001b[32m234\u001b[39m         chunks = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_reader\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_low_memory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    235\u001b[39m         \u001b[38;5;66;03m# destructive to chunks\u001b[39;00m\n\u001b[32m    236\u001b[39m         data = _concatenate_chunks(chunks)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:838\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader.read_low_memory\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:905\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._read_rows\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:874\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:891\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._check_tokenize_status\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mparsers.pyx:2061\u001b[39m, in \u001b[36mpandas._libs.parsers.raise_parser_error\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[31mParserError\u001b[39m: Error tokenizing data. C error: Expected 1 fields in line 18, saw 2\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "import io\n",
    "\n",
    "# Download immune subtypes data from TCGA Pan-Cancer Atlas\n",
    "url = \"https://api.gdc.cancer.gov/data/1a7d7be8-675d-4e60-a105-19d4121bdebf\"\n",
    "headers = {\n",
    "    'Content-Type': 'application/json',\n",
    "}\n",
    "\n",
    "response = requests.get(url, headers=headers)\n",
    "if response.status_code == 200:\n",
    "    # Read the CSV file\n",
    "    subtypes = pd.read_csv(io.StringIO(response.text))\n",
    "    print(\"Successfully downloaded immune subtypes data\")\n",
    "    print(\"Data shape:\", subtypes.shape)\n",
    "    print(\"Columns:\", subtypes.columns)\n",
    "    \n",
    "    # Filter for immune subtype information\n",
    "    if 'Immune.Subtype' in subtypes.columns:\n",
    "        subtypes = subtypes[['TCGA.Participant.Barcode', 'Immune.Subtype']]\n",
    "        subtypes.columns = ['SampleID', 'ImmuneSubtype']\n",
    "        \n",
    "        # Merge with expression data\n",
    "        expr_df = expr_df.join(subtypes.set_index('SampleID'), on='sampleId')\n",
    "        expr_df.dropna(subset=['ImmuneSubtype'], inplace=True)\n",
    "        print(\"Merged data shape:\", expr_df.shape)\n",
    "    else:\n",
    "        print(\"Could not find immune subtype column in the data\")\n",
    "else:\n",
    "    print(\"Could not download immune subtypes data. Status code:\", response.status_code)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc61ae18",
   "metadata": {},
   "source": [
    "## ðŸ©º Fetch Real Clinical Survival Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "08426b39",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of [Index(['OS_MONTHS', 'OS_STATUS'], dtype='object')] are in the [columns]\"",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[36]\u001b[39m\u001b[32m, line 5\u001b[39m\n\u001b[32m      3\u001b[39m clinical_df = pd.DataFrame(clinical_data)\n\u001b[32m      4\u001b[39m clinical_df = clinical_df.set_index(\u001b[33m'\u001b[39m\u001b[33mpatientId\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m----> \u001b[39m\u001b[32m5\u001b[39m clinical_df = \u001b[43mclinical_df\u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mOS_MONTHS\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mOS_STATUS\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m.rename(columns={\n\u001b[32m      6\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mOS_MONTHS\u001b[39m\u001b[33m'\u001b[39m: \u001b[33m'\u001b[39m\u001b[33mOS_Time\u001b[39m\u001b[33m'\u001b[39m,\n\u001b[32m      7\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mOS_STATUS\u001b[39m\u001b[33m'\u001b[39m: \u001b[33m'\u001b[39m\u001b[33mOS_Status\u001b[39m\u001b[33m'\u001b[39m\n\u001b[32m      8\u001b[39m })\n\u001b[32m     10\u001b[39m \u001b[38;5;66;03m# Format status to 1/0\u001b[39;00m\n\u001b[32m     11\u001b[39m clinical_df[\u001b[33m'\u001b[39m\u001b[33mOS_Status\u001b[39m\u001b[33m'\u001b[39m] = clinical_df[\u001b[33m'\u001b[39m\u001b[33mOS_Status\u001b[39m\u001b[33m'\u001b[39m].apply(\u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[32m1\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mstr\u001b[39m(x).strip().upper().startswith(\u001b[33m'\u001b[39m\u001b[33mDECEASED\u001b[39m\u001b[33m'\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m \u001b[32m0\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.12.4/lib/python3.12/site-packages/pandas/core/frame.py:4108\u001b[39m, in \u001b[36mDataFrame.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4106\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[32m   4107\u001b[39m         key = \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[32m-> \u001b[39m\u001b[32m4108\u001b[39m     indexer = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m.\u001b[49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mcolumns\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[32m1\u001b[39m]\n\u001b[32m   4110\u001b[39m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[32m   4111\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[33m\"\u001b[39m\u001b[33mdtype\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) == \u001b[38;5;28mbool\u001b[39m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.12.4/lib/python3.12/site-packages/pandas/core/indexes/base.py:6200\u001b[39m, in \u001b[36mIndex._get_indexer_strict\u001b[39m\u001b[34m(self, key, axis_name)\u001b[39m\n\u001b[32m   6197\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   6198\u001b[39m     keyarr, indexer, new_indexer = \u001b[38;5;28mself\u001b[39m._reindex_non_unique(keyarr)\n\u001b[32m-> \u001b[39m\u001b[32m6200\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   6202\u001b[39m keyarr = \u001b[38;5;28mself\u001b[39m.take(indexer)\n\u001b[32m   6203\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[32m   6204\u001b[39m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.12.4/lib/python3.12/site-packages/pandas/core/indexes/base.py:6249\u001b[39m, in \u001b[36mIndex._raise_if_missing\u001b[39m\u001b[34m(self, key, indexer, axis_name)\u001b[39m\n\u001b[32m   6247\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m nmissing:\n\u001b[32m   6248\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m nmissing == \u001b[38;5;28mlen\u001b[39m(indexer):\n\u001b[32m-> \u001b[39m\u001b[32m6249\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m]\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m   6251\u001b[39m     not_found = \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask.nonzero()[\u001b[32m0\u001b[39m]].unique())\n\u001b[32m   6252\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m not in index\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[31mKeyError\u001b[39m: \"None of [Index(['OS_MONTHS', 'OS_STATUS'], dtype='object')] are in the [columns]\""
     ]
    }
   ],
   "source": [
    "# Get clinical data\n",
    "clinical_data = get_cbioportal_data(f'studies/{study_id}/clinical-data')\n",
    "clinical_df = pd.DataFrame(clinical_data)\n",
    "clinical_df = clinical_df.set_index('patientId')\n",
    "clinical_df = clinical_df[['OS_MONTHS', 'OS_STATUS']].rename(columns={\n",
    "    'OS_MONTHS': 'OS_Time',\n",
    "    'OS_STATUS': 'OS_Status'\n",
    "})\n",
    "\n",
    "# Format status to 1/0\n",
    "clinical_df['OS_Status'] = clinical_df['OS_Status'].apply(lambda x: 1 if str(x).strip().upper().startswith('DECEASED') else 0)\n",
    "\n",
    "# Map patient ID from sample ID (remove '_tumor' or use substring)\n",
    "expr_df['patientId'] = expr_df.index.str.replace('-Tumor', '', regex=False).str.slice(0, 12)\n",
    "expr_df = expr_df.join(clinical_df, on='patientId')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adbc3e15",
   "metadata": {},
   "source": [
    "## ðŸ§ª Survival Analysis by CD8A Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679ec5c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from lifelines import KaplanMeierFitter\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "merged = expr_df.dropna(subset=['OS_Time', 'OS_Status'])\n",
    "median_cd8 = merged['CD8A'].median()\n",
    "high = merged[merged['CD8A'] > median_cd8]\n",
    "low = merged[merged['CD8A'] <= median_cd8]\n",
    "\n",
    "kmf = KaplanMeierFitter()\n",
    "kmf.fit(high['OS_Time'], high['OS_Status'], label='High CD8A')\n",
    "ax = kmf.plot()\n",
    "kmf.fit(low['OS_Time'], low['OS_Status'], label='Low CD8A')\n",
    "kmf.plot(ax=ax)\n",
    "plt.title(\"Survival by CD8A Expression\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
